### ISSUE TITLE
newline-before-return: bug with comment

### ISSUE THREAD

**Issue by anton-rudeshko** [eslint/eslint#5480]
**What version are you using?**
```
$ eslint --version
v2.3.0
```

**What did you do?**
.eslintrc.js:

```js
module.exports = {
    rules: {
        'newline-before-return': 'error'
    }
};
```

okay.js:

```js
function main() {
    var test = {
    };

    return 1;
}
```

error.js:

```js
function main() {
    var test = {
        //
    };

    return 1;
}
```

**What happened?**
```
$ eslint .

error.js
  6:5  error  Expected newline before return statement  newline-before-return

✖ 1 problem (1 error, 0 warnings)
```

**What did you expect to happen?**
Should not be considered problem

**Comment by btmills** [comment]
Just tried this and confirmed. That's weird...

**Comment by btmills** [comment]
It looks like there's a problem with how [comments are collected and calculated](https://github.com/eslint/eslint/blob/ef341be49669230bb33c76886c666be89c8e99ee/lib/rules/newline-before-return.js#L107-L109). @anton-rudeshko are you interested in submitting a PR?

**Comment by anton-rudeshko** [comment]
Yeah, but not sure yet if I can find time for that. Will try. At least it known bug now.

**Comment by diki** [comment]
Looks like problem occurs because leading comments of `return` node is miscalculated https://github.com/eslint/eslint/blob/master/lib/rules/newline-before-return.js#L70

**fail**
```
function main() {
  var test = {
    //
  };

  //
  return 1;
}

calcCommentLines = 2
```

**also fail**
```
function main() {
  var test = {
    //
   /*

   */
  };

  //
  return 1;
}

calcCommentLines = 5
```
**pass**
```
function main() {
  var test = {
    //
  };
  var foo = 5;

  //
  return 1;
}
calcCommentLines = 1
```

It is because `sourceCode.getComments(node).leading` adds comment lines inside object expression if there are no other expressions between return and object. Comparing `comment.loc.start.line` and `tokenBefore.loc.end.line` and adding only comment lines bigger than tokenBefore line https://github.com/diki/eslint/commit/1dcd897ad64fb4df5ae8e43b28ebf8ecff618f77 fixes the issue.

**Comment by diki** [comment]
@anton-rudeshko @btmills do you want me to send a pul request for this.

**Comment by anton-rudeshko** [comment]
@diki I don't mind ) Make the world a better place if you can ;)

**Comment by diki** [comment]
@anton-rudeshko :), sent PR

**Comment by kaicataldo** [comment]
The proposed solution in the PR looks good to me, though I can't help but wonder why comments inside the previous token are getting returned at all. It only seems to be an issue when the previous object/block only contain a comment and nothing else. Granted, I don't see the point of having an empty block or object literal with a comment inside of it, but is that the expected behavior of 'sourceCode.getComments(node).leading'?

If not, I can make an issue and look into fixing that.
